### DISCUSSION TITLE
How to integrate an Arduino library into tasmota ?

### DISCUSSION THREAD

**Discussion by darcyg** [arendst/Tasmota#16218] [Q&A]
I need to use modbus functionality in esp32.
Tasmota's modbus only supports esp8266.
I am trying to port arduino's modbus library to tasmota.
I found two libraries
https://github.com/eModbus/eModbus
https://github.com/arduino-libraries/ArduinoModbus
But I can't get tasmota to compile both libraries
I put the modbus library in the lib/lib_modbus directory
and modify lib_extra_dirs configured by tasmota32base env

Can you tell me if you want to add a third-party library
like modbus or https://github.com/espressif/esp-modbus

How should it be integrated into tasmota

I am new to arduino development. Not sure how to start.
thanks

**Comment by barbudor** [comment]
Tasmota already has a TasmotaModbus library which supports our different usage (various drivers, script for SML,  recent MODBUS gateway driver) 

If this doesn't work with ESP32 then the 1st thing to do would be to make it work before trying to add an external dependency

**Reply by arendst** [reply]
Let me chk the modbus interface between an ESP32 and a SDM230 using modbus interface which I expect to work....

**Reply by arendst** [reply]
As expected. No problem at all:

![image](https://user-images.githubusercontent.com/11044339/184496935-0e301188-5b72-4d05-8a14-f16e1d5c53ba.png)

**Reply by darcyg** [reply]
Tested the modbus of tasmota
many problems
Read Coils and Read Discrete Inputs
Unable to return data
Read Multiple Holding Registers and Read Input Registers can return normally
But count cannot be set > 9. No return if more than 9

**Reply by barbudor** [reply]
@darcyg in any case,  the priority is to solve/improve those possible problems with the tasmota libray
We will not accept a redundant third party library when we already have native one that should cover Tasmota's MODBUS usage we need

**Reply by darcyg** [reply]
cmnd command

> ModbusBaudrate 9600
> ModbusSerialConfig 3
> ModbusSend {"deviceAddress":1,"functionCode":1,"startAddress":1,"type":"bit","count":1}

no return data

> ModbusSend {"deviceAddress":1,"functionCode":3,"startAddress":1,"type":"uint16","count":8}

return data

> {"ModbusReceived":{"DeviceAddress":1,"FunctionCode":3,"StartAddress":1,"Length":21,"Count":8,"Values":[0,0,0,0,0,0,0,0]}}

in berry script

> def slog(val)
>   var len=size(val)
>   print(len," | ",val)
> end
> ser=serial(16, 4, 9600, serial.SERIAL_8N1)
> 
> ser.write(bytes().fromhex("01020000000879CC"))
> tasmota.delay(1000)
> msg=ser.read()
> slog(msg.tohex())
> slog(msg.asstring())

**return data:**

> 12 | 01020100A188

**Comment by darcyg** [comment]
Thank you for your response.

I will test the official MODBUS library.

In addition, I still need to integrate the code of other third-party CPPs.

I don't know how to make tasmota compile the third-party Arduino library or CPP code

Can you give me a tutorial or tell me where the learning materials are?

**Reply by barbudor** [reply]
There are some docs about Tasmota architecture here:  https://tasmota.github.io/docs/For-Developers/#editing-articles

For libraries, you should follow the way other libraries are integrated in one of the libraries package (folder)

**Reply by darcyg** [reply]
I have tried integrating several arduino libraries into tasmota. But unsuccessful. I tried platformio's method of importing the arduino library. Neither was successful.
I don't know how to create a library package that can have tasmota import compile. In other embedded systems I can find build scripts like makefile/cmake. Just add the source code I want to compile. But I can't find these in tasmota.

**Reply by barbudor** [reply]
@darcyg it's only about adding a folder in one of the predefined package such as lib_div...

**Reply by darcyg** [reply]
@barbudor I tried it. I can see scanning dependencies
115 compatible libraries found.
But not in Dependency Graph
I found that the library I added cannot be compiled
I don't know if I have to write the ino of xdrv
In order for the library to be compiled into the Dependency Graph

**Reply by barbudor** [reply]
Adding a library itself of course does nothing if you don't use it inside tasmota core or a driver...
But at least it should still be compiled

Anyhow without any detail, your help is limited to Crystal Ball guessing

**Reply by darcyg** [reply]
Yes, it should be added to compile.
The problem I'm having is that the added code doesn't get compiled.
Unable to generate .a library file
I don't know where the problem is. I modify the parameters of several ini files of platformio. No effect.
How do I need to provide details? 
I understand. in a clean tasmota branch.
In the lib/lib_basic directory. Add a third-party library for arduino
When building, the relevant .o and .a files should be generated in the .pio build directory.
But I can't see the .o and .a files of the new library anyway.. I don't know what's wrong
Could not find tasmota's makefile or cmake such files. So it's not clear to me what's wrong with the build.

**Reply by Jason2866** [reply]
There has to be a `library.properties` in the folder where you added your lib. If you provide a link to the fork of Tasmota where you have added, we will have a look. But as barbudor wrote nothing will work as long there is no Tasmota driver which calls the lib.

**Comment by nikito7** [comment]
https://github.com/phpcoder/tasmota-custom-integration
